core/src/main/scala/org/apache/spark/status/api/v1/OneApplicationResource.scala

```diff
@@ -51,6 +52,43 @@ private[v1] class AbstractApplicationResource extends BaseAppResource {
   @Path("executors")
   def executorList(): Seq[ExecutorSummary] = withUI(_.store.executorList(true))
 
+  @GET
+  @Path("executors/{executorId}/threads")
+  def threadDump(@PathParam("executorId") executorId: String): Response = withUI { ui =>
+    uiRoot.executorThreadDumpsNotAvailableError().getOrElse {
+      val safeExecutorId =
+        Option(UIUtils.stripXSS(executorId)).map { executorId =>
+          UIUtils.decodeURLParameter(executorId)
+        }.getOrElse {
+          throw new IllegalArgumentException(s"Missing executorId parameter")
+        }
+
+      def isAllDigits(x: String) = x.forall(Character.isDigit)
+
+      if (executorId != SparkContext.DRIVER_IDENTIFIER && !isAllDigits(executorId)) {
+        Response.serverError()
+          .entity(s"Invalid executorId: neither '${SparkContext.DRIVER_IDENTIFIER}' nor number.")
+          .status(Response.Status.BAD_REQUEST)
+          .build()
+      } else {
+        if (ui.store.asOption(ui.store.executorSummary(executorId)).exists(!_.isActive)) {
```

Comment:
There is a minor race here in which the active state can change in between here and when you get the thread dump below. Can you handle it in the call below instead (e.g. if it throws an exception when the executor is stopped)?

Reply:
In that case I would suggest to consider changing the log level in org.apache.spark.SparkContext#getExecutorThreadDump

```  scala
        logError(s"Exception getting thread dump from executor $executorId", e)
```

Maybe to info. What is your opinion?

Reply:
Since that method doesn't throw an exception, perhaps it's not worth it to try to fix this race nor make the log change.

Also I think the logic here is a little off. If the executor doesn't exist at all in the store, this condition will not trigger.

```
scala> case class Exec(active: Boolean)
scala> Option[Exec](null).exists(!_.active)
res3: Boolean = false
```

In that situation you should probably return a 404 error (= throw NotFoundException).

Reply:
This is exactly the current behaviour as for false it goes into:

```scala
} else {
          ui.sc.flatMap(_.getExecutorThreadDump(safeExecutorId))
            .map(Response.ok(_).build())
            .getOrElse(
              Response.serverError()
                .entity("No stack traces are available.")
                .status(Response.Status.NOT_FOUND)
                .build())
        }
```

Then further to:

```scala
 .getOrElse(
              Response.serverError()
                .entity("No stack traces are available.")
                .status(Response.Status.NOT_FOUND)
                .build())
```

See the test screenshot for "Not existing (well formatted) executor ID". 

Or did I misunderstood your requirement?

Reply:
Those two error messages say "No stack traces are available", while the actual error I'm pointing out is that the executor may not exist at all. The HTTP error code is the same, it's just that the message is misleading.

e.g. if you have 5 executors and ask for the stack trace from executor 1000. Yeah, technically, "No stack traces are available" is correct, it's just that we can do better.

Reply:
Yeah, I see your point.